### ISSUE TITLE
TPointerMap.Delete can have undefined result + off by one error

### ISSUE THREAD

**Issue by JBontes** [graphics32/graphics32#14]
The code for TPointerMap.Delete is incorrect:

```
function TPointerMap.Delete(BucketIndex, ItemIndex: Integer): PData;
begin
  with FBuckets[BucketIndex] do
  begin
    Result := Items[ItemIndex].Data;

    if FCount = 0 then Exit;   <<-- error: how can result be valid if count = 0?

    Dec(Count);
    if Count = 0 then
      SetLength(Items, 0)
    else
    if (ItemIndex < Count) then
       //Oops off by 1 error!
      Move(Items[ItemIndex + 1], Items[ItemIndex], (Count - ItemIndex - 1) * SizeOf(TPointerBucketItem));
  end;
  Dec(FCount); <<-- The use of with makes this statement confusing.
end;
```

The code should be changed as follows:
```
{
  TPointerBucket = record
    Count: Integer;
    Items: TPointerBucketItemArray;
end;}

function TPointerMap.Delete(BucketIndex, ItemIndex: Integer): PData;
var
  Bucket: TPointerBucket ;
begin
    if FCount = 0 then Exit(nil);
    //Perhaps add some code to validate BucketIndex & ItemIndex?
    Assert(BucketIndex < Length(FBuckets));
    Bucket:= FBuckets[BucketIndex];
    if ItemIndex >= Bucket.
    Assert(ItemIndex < Length(Bucket.Items));
    Result := Bucket.Items[ItemIndex].Data;
    Dec(Bucket.Count);
    if Bucket.Count = 0 then
      SetLength(Bucket.Items, 0)
    else
     /// assume array like so: 0 1 2 3 4  , itemindex = 0
    /// result should be 1 2 3 4
    /// move(1,0,4) (because 4 items should be moved.
    /// Thus move (itemindex+1, intemindex, count-itemindex)
    if (ItemIndex < Bucket.Count) then
      Move(Items[ItemIndex + 1], Items[ItemIndex], (Bucket.Count - ItemIndex) * SizeOf(TPointerBucketItem));
  end;
  Dec(FCount);
end;
```

**Comment by fernando-takeshi** [comment]
I've encountered a problem that @JBontes' observation about the 'one off' proved to be accurate, but instead of using his altered code, I just removed the "-1" from the Move count parameter.

@JBontes: Your comments about validating the indexes and such are understandable, but reading the code it seems to me that the classes that use / extend TPointerMap are responsible for taking care of memory allocation/deallocation, thus the implementation of TPointerMap.Delete becomes a bit clearer. Consider the following code implemented on GR32_MicroTiles.TMicroTilesMap.Delete, for example. In any case, thanks for opening this issue - it has helped me solve a problem we were having. Cheers!
